### PR TITLE
feat: add utility function to calculate time before issue deadline

### LINKED ISSUE #40: #13: Add functionality to check issue's deadline
1) Read guideline [here](https://gist.github.com/iamnovichek/801ac395a06a7d8b04b022084ace684d)
2) Create function `get_time_before_deadline`  in `tracker/utils.py` to check how much time it rests till deadline
3) Fetch the info about the specific repository and compute the time till deadline will be missed for each issue only if it has an assignee
4) Provide some screenshots to prove that your code works

### PR DESCRIPTION
**Pull request by Jagadeeshftw**
closes #40 

## Testing:
```
[root@puppet ~]# python3 testing.py

=== Test Results ===
Test Case: Deadline in 2 days (default case with days remaining)
Result: 1 days remaining
--------------------------------------------------
Test Case: Deadline in 2 hours (case with hours remaining)
Result: 0 hours 59 minutes remaining
--------------------------------------------------
Test Case: Deadline passed immediately (deadline passed case)
Result: Deadline has passed.
--------------------------------------------------
Test Case: Repository not found (invalid repository)
Result: Repository not found.
--------------------------------------------------
[root@puppet ~]# cat testing.py
from datetime import datetime, timedelta, timezone
from unittest.mock import Mock, patch

class MockRepository:
    def __init__(self, author, name, time_limit):
        self.author = author
        self.name = name
        self.time_limit = time_limit

def mock_get_repository(author, name):
    mock_data = {
        ("test_author", "test_repo"): MockRepository("test_author", "test_repo", 172800),  # 2 days in seconds
        ("test_author", "short_deadline_repo"): MockRepository("test_author", "short_deadline_repo", 7200),  # 2 hours
        ("test_author", "past_deadline_repo"): MockRepository("test_author", "past_deadline_repo", 0),  # Immediate deadline
    }
    return mock_data.get((author, name))

def check_issue_assignment_events(issue):
    return {
        "assignee": "test_user",
        "assigned_at": (datetime.utcnow() - timedelta(hours=1)).strftime("%Y-%m-%dT%H:%M:%SZ"),  # Assigned 1 hour ago
    }

def get_time_before_deadline(issue: dict) -> str:
    """
    Returns the time remaining before the deadline of an assigned issue.
    If the issue has no assignee or deadline, returns appropriate messages.
    """
    assignment_info = check_issue_assignment_events(issue)
    assigned_at = assignment_info.get("assigned_at")

    if not assigned_at:
        return "This issue is not assigned."

    repository_details = issue.get("repository_details")
    if not repository_details:
        return "Repository details not found."

    repo = mock_get_repository(repository_details["author"], repository_details["name"])
    if not repo:
        return "Repository not found."

    assigned_time = datetime.strptime(assigned_at, "%Y-%m-%dT%H:%M:%SZ").replace(tzinfo=timezone.utc)
    deadline_datetime = assigned_time + timedelta(seconds=repo.time_limit)
    now = datetime.now(timezone.utc)

    if deadline_datetime > now:
        remaining_time = deadline_datetime - now
        if remaining_time.days > 0:
            return f"{remaining_time.days} days remaining"
        else:
            hours = remaining_time.seconds // 3600
            minutes = (remaining_time.seconds % 3600) // 60
            return f"{hours} hours {minutes} minutes remaining"
    else:
        return "Deadline has passed."

test_cases = [
    {
        "description": "Deadline in 2 days (default case with days remaining)",
        "repository_details": {"author": "test_author", "name": "test_repo"},
    },
    {
        "description": "Deadline in 2 hours (case with hours remaining)",
        "repository_details": {"author": "test_author", "name": "short_deadline_repo"},
    },
    {
        "description": "Deadline passed immediately (deadline passed case)",
        "repository_details": {"author": "test_author", "name": "past_deadline_repo"},
    },
    {
        "description": "Repository not found (invalid repository)",
        "repository_details": {"author": "unknown_author", "name": "unknown_repo"},
    },
]

print("\n=== Test Results ===")
for case in test_cases:
    issue = {
        "repository_details": case["repository_details"],
    }
    print(f"Test Case: {case['description']}")
    print("Result:", get_time_before_deadline(issue))
    print("-" * 50)

[root@puppet ~]#

### CODE REVIEW

**Comment by Jagadeeshftw** [comment]
Can you review this @iamnovichek?

File: tracker/utils.py [hunk 1]
```diff
@@ -317,3 +322,39 @@ def attach_link_to_issue(issue_title: str, issue_link: str) -> str:
...
+    deadline_str = issue.get("due_on")
+    if not deadline_str:
+        return "No deadline set for this issue."
+
+    deadline_datetime = datetime.strptime(deadline_str, "%Y-%m-%dT%H:%M:%SZ").replace(
+        tzinfo=timezone.utc
+    )
+    now = datetime.now(timezone.utc)
+
+    time_left = deadline_datetime - now
```
Comment (iamnovichek) [inline tracker/utils.py]: Pls use `relativedelta`

**Review by iamnovichek — CHANGES_REQUESTED** [review]

**Comment by iamnovichek** [comment]
@Jagadeeshftw Code review done! Pls apply changes

**Comment by iamnovichek** [comment]
@Jagadeeshftw pls sync with main

File: tracker/utils.py (hunk 1 above)
Comment (Jagadeeshftw) [reply tracker/utils.py]: Updated!

File: tracker/utils.py [hunk 2]
```diff
@@ -317,3 +322,40 @@ def attach_link_to_issue(issue_title: str, issue_link: str) -> str:
...
+    :return: Time remaining in a human-readable format.
+    """
+
+    assignment_info = check_issue_assignment_events(issue)
+    assigned_at = assignment_info.get("assigned_at")
+
+    if not assigned_at:
+        return "This issue is not assigned."
+
+    deadline_str = issue.get("due_on")
```
Comment (iamnovichek) [inline tracker/utils.py]: No, the original GH issue object doesn't have this field. You should calculate it yourself.

Here you have `time_limit` filed (in hours).
https://github.com/KostCode-Organization/hackathon_bot/blob/main/tracker/models.py#L158

You should just check what repository the issue if related to -> get the corresponding `Repository` object and finally calculate the time till deadline based on `time_limit` value

**Review by iamnovichek — CHANGES_REQUESTED** [review]

**Comment by iamnovichek** [comment]
@Jagadeeshftw Code review done! Pls apply changes

**Comment by iamnovichek** [comment]
@Jagadeeshftw pls sync with main

File: tracker/utils.py [hunk 3]
```diff
@@ -317,3 +322,49 @@ def attach_link_to_issue(issue_title: str, issue_link: str) -> str:
...
+
+    assigned_time = datetime.strptime(assigned_at, "%Y-%m-%dT%H:%M:%SZ").replace(
+        tzinfo=timezone.utc
+    )
+    deadline_datetime = assigned_time + timedelta(seconds=time_limit_seconds)
+    now = datetime.now(timezone.utc)
+
+    if deadline_datetime > now:
+        remaining_time = deadline_datetime - now
+        return f"Time remaining: {remaining_time.days} days, {remaining_time.seconds // 3600} hours"
```
Comment (iamnovichek) [inline tracker/utils.py]: Move `3600` magic number to `values.py` and import it from there

File: tracker/utils.py [hunk 4]
```diff
@@ -317,3 +322,49 @@ def attach_link_to_issue(issue_title: str, issue_link: str) -> str:
...
+        return "Repository details not found."
+
+    from .models import Repository
+
+    repo = Repository.objects.get(
+        author=repository_details["author"], name=repository_details["name"]
+    )
+    time_limit_seconds = repo.time_limit
+
+    assigned_time = datetime.strptime(assigned_at, "%Y-%m-%dT%H:%M:%SZ").replace(
```
Comment (iamnovichek) [inline tracker/utils.py]: Move this datetime format to `values.py` and import it from there

File: tracker/utils.py [hunk 5]
```diff
@@ -317,3 +322,49 @@ def attach_link_to_issue(issue_title: str, issue_link: str) -> str:
...
+        return "This issue is not assigned."
+
+    repository_details = get_repository_from_issue(issue)
+    if not repository_details:
+        return "Repository details not found."
+
+    from .models import Repository
+
+    repo = Repository.objects.get(
+        author=repository_details["author"], name=repository_details["name"]
```
Comment (iamnovichek) [inline tracker/utils.py]: Pls use `.get()` instead of `[]`

**Review by iamnovichek — CHANGES_REQUESTED** [review]
Pls provide some screenshots to prove that your code works

File: tracker/utils.py (hunk 5 above)
Comment (Jagadeeshftw) [reply tracker/utils.py]: Updated!

File: tracker/utils.py (hunk 4 above)
Comment (Jagadeeshftw) [reply tracker/utils.py]: Updated!

File: tracker/utils.py (hunk 3 above)
Comment (Jagadeeshftw) [reply tracker/utils.py]: Updated!

**Review by iamnovichek — APPROVED** [review]

**Pull request merged**